### PR TITLE
Fixed invalid value updates of dependent values in LoopUnrolling

### PR DESCRIPTION
**Pull request by m4rs-mt**
This PR is part of the solution for issue report #652. As also described in PR #657, the following kernel program raises an exception in the `Release` and `O2` pipelines using the latest `master` branch:
```c#
            int value = 3;
            int counter = 0;
            do
            {
                ++value;
            }
            while (counter++ < 3);
            data[index] = value;
```
This PR fixes the `LoopUnrolling` transformation by tracking value updates for values being referenced outside of the loop in the presence of `do-while` loops. It also adds an "additional" test cases (it is an old test case which accidentally did not cover this edge case.... 😞) to reproduce this problem.

However, this PR does not fix all the problems related to "do-while" loops (yet...), since we have to perform an additional unwinding step at the beginning (+1 additional "iteration" compared to "while" loops). This raises the question of how to distinguish between these two types of loops. One possibility would be to update PR #657 to adjust the trip count of these loops.

@MoFtZ @deng0: What do you think?

*Note that this PR is a draft and not ready to be used in production*

### CODE REVIEW

**Comment by deng0** [comment]
@m4rs-mt I made some more tests and could not find any additional problems.
Only the ones I mentioned in my initial pull request:

open issues:
- critical: do not unroll when iter value is modified inside the loop
- optional: add support for ushort, short, byte, sbyte iter types

**Comment by jgiannuzzi** [comment]
@m4rs-mt I agree that the support for additional iter types should be part of `v1.1` and not `v1.0`. We are too late in the release process to introduce such changes and should only focus on fixing critical bugs.

Thanks a lot @deng0 for reporting these issues, and thanks a lot @m4rs-mt for fixing this!

PS: this is also how I feel 😅:
> @m4rs-mt Thank you for your work. I don't understand most of it.

**Comment by deng0** [comment]
@m4rs-mt I agree, the other iter types can wait for v1.1.

I've just checked all my testkernels with your latest commit and now they all produce the correct results.
Great work!

File: Src/ILGPU/IR/Transformations/LoopUnrolling.cs [hunk 1]
```diff
@@ -465,6 +519,10 @@ private struct LoopProcessor : Loops.ILoopProcessor
                 // Replace all phi values with their computed values
                 loopSpecializer.ReplacePhis();
 
+                // Replace all values with used outside of the loop with their computed
+                // values
```
Comment (MoFtZ) [inline Src/ILGPU/IR/Transformations/LoopUnrolling.cs:523]: MINOR: Spelling.
```suggestion
                // Replace all values used outside of the loop with their computed values
```

**Review by MoFtZ — APPROVED** [review]
